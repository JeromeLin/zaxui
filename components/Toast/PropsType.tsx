interface BaseProps {
  visible?: boolean;
  stayTime?: number;
  onMaskClick?: (x?: any) => void;
}

export interface ToastProps extends BaseProps {
  prefixCls?: string;
  className?: string;
}
