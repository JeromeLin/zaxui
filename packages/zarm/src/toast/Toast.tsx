import React, { useCallback, useEffect, useRef, useState } from 'react';
import BaseToastProps from './interface';
import Popup from '../popup';
import { ConfigContext } from '../n-config-provider';

export interface ToastProps extends BaseToastProps {
  className?: string;
}
export interface UseToast {
  show: (propsType: ToastProps) => void;
  hide: () => void;
}
interface CompoundedComponent
  extends React.ForwardRefExoticComponent<ToastProps & React.RefAttributes<HTMLElement>> {
  useToast: () => UseToast;
}
const Toast = React.forwardRef<unknown, ToastProps>((props, ref) => {
  const { prefixCls: globalPrefixCls } = React.useContext(ConfigContext);
  const prefixCls = `${globalPrefixCls}-toast`;

  const toastRef = (ref as any) || React.createRef<HTMLElement>();
  const { className, stayTime, content, visible: propVisible, afterClose, ...others } = props;

  const [visible, setVisible] = useState(propVisible!);

  const timerRef = useRef<number>();

  const _hide = () => {
    setVisible(false);
  };

  const autoClose = useCallback(() => {
    if ((stayTime as number) > 0) {
      timerRef.current = window.setTimeout(() => {
        _hide();
        clearTimeout(timerRef.current);
      }, stayTime);
    }
  }, [stayTime]);

  useEffect(() => {
    autoClose();
    return () => {
      clearTimeout(timerRef.current);
    };
  }, [autoClose]);

  return (
    <Popup
      direction="center"
      maskType="transparent"
      width="70%"
      {...others}
      visible={visible}
      afterClose={afterClose}
    >
      <div className={prefixCls} ref={toastRef}>
        <div className={`${prefixCls}__container`}>{content}</div>
      </div>
    </Popup>
  );
}) as CompoundedComponent;

Toast.displayName = 'Toast';

Toast.defaultProps = {
  visible: false,
  stayTime: 3000,
  mask: false,
};

export default Toast;
