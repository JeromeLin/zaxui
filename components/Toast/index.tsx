import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';
import Mask from '../Mask';

export interface ToastProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Toast extends PureComponent<ToastProps, any> {

  static defaultProps = {
    prefixCls: 'za-toast',
    visible: false,
    stayTime: 3000,
  };

  private timer: number;

  constructor(props) {
    super(props);
    this.state = {
      isShow: false,
    };
  }

  componentDidMount() {
    if (this.props.visible) {
      this.enter(this.props.stayTime, this.props.onMaskClick);
    }
  }

  componentWillReceiveProps(nextProps) {
    clearTimeout(this.timer);

    if (nextProps.visible) {
      this.enter(nextProps.stayTime, nextProps.onMaskClick);
    } else {
      this.leave();
    }
  }

  componentWillUnmount() {
    clearTimeout(this.timer);
  }

  enter = (stayTime, onMaskClick) => {
    this.setState({
      isShow: true,
    });

    if (stayTime === 0) {
      return;
    }

    this.timer = setTimeout(() => {
      if (typeof onMaskClick === 'function') {
        onMaskClick();
      }
      clearTimeout(this.timer);
    }, stayTime);
  }

  leave = () => {
    this.setState({
      isShow: false,
    });
  }

  render() {
    const { prefixCls, className, visible, children, onMaskClick } = this.props;

    const cls = classnames(`${prefixCls}`, className, {
      [`${prefixCls}-open`]: this.state.isShow,
    });

    return (
      <div className={cls}>
        <div className={`${prefixCls}-container`}>
          {children}
        </div>
        <Mask visible={visible} type="transparent" onClose={onMaskClick} />
      </div>
    );
  }
}
