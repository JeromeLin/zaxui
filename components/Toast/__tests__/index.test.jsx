import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Toast from '../index';

describe('Toast', () => {
  it('renders correctly', () => {
    const onMaskClick = jest.fn();

    const wrapper = shallow(
      <Toast onMaskClick={onMaskClick}>foo</Toast>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.setProps({ visible: true });
    wrapper.unmount();
  });

  it('duration', () => {
    const wrapper = shallow(
      <Toast duration={0}>foo</Toast>
    );
    wrapper.setProps({ visible: true });
  });
});
