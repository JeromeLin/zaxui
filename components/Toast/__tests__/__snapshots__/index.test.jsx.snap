// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast renders correctly 1`] = `
<div
  className="za-toast"
>
  <div
    className="za-toast-container"
  >
    foo
  </div>
  <Mask
    className={null}
    onClose={[Function]}
    prefixCls="za-mask"
    type="transparent"
    visible={false}
  />
</div>
`;
